### PR TITLE
DropdownMenu objects

### LINKED ISSUE #17: resolve conflicts between sidebar and dropdown (shiny.semantic)

### LINKED ISSUE #38: Add task_item, notification_item, message_item functions

### LINKED ISSUE #39: Add dropdown_menu, dropdown_menu_output and render_dropdown_menu functions

### LINKED ISSUE #40: Add info_box_output and render_info_box functions

### LINKED ISSUE #41: Add info_box function ( for now same as value_box)

### PR DESCRIPTION
**Pull request by maju116**
Added `dropdown_menu`, `message_item`,` task_item` function with aliases. 
Added `infoBox`, `infoBoxOutput`, `renderInfoBox` function (as aliases for `value_box` class functions)
JS changed for `sidebar and pusher`.

Closes #38 
Closes #39 
Closes #17 
Closes #40 
Closes #41 

3t

### CODE REVIEW

File: tests/testthat/test-names.R [hunk 1]
```diff
@@ -0,0 +1,12 @@
+library(semantic.dashboard)
+library(shinydashboard)
+context("Function names")
+
+test_that("Fucntion names from shinydashboard are identical to names from semantic.dashboard", {
```
Comment (marekrogala) [inline tests/testthat/test-names.R]: hmm I'd say the _essence_ of this test is less `Fucntion names from shinydashboard are identical to names from semantic.dashboard` and more something like `All functions from shinydashboard are implemented except functions known to be missing.`

File: R/semantic_dashboard.R [hunk 2]
```diff
@@ -6,8 +6,9 @@
 #' @export
 dashboard_header <- function(..., color = "black"){
   verify_value_allowed("color", ALLOWED_COLORS)
-  shiny::div(class = paste("ui top attached inverted", color, "demo menu"),
-             shiny::tags$a(class = "item", shiny::tags$i(class = "sidebar icon"), "Menu"), ...)
+  shiny::div(class = paste("ui top attached inverted", color, " menu"),
+             shiny::tags$a(id = "toggle_menu", class = "item", shiny::tags$i(class = "sidebar icon"), "Menu"),
+             shiny::tags$div(style = "margin-left:auto", class = paste("ui icon buttons", color), ...))
```
Comment (marekrogala) [inline R/semantic_dashboard.R]: nit: missing space after `marginleft:`

File: R/uidropdown_menu.R [hunk 3]
```diff
@@ -0,0 +1,82 @@
+#' Create a dropdown menu (alias for \code{dropdown_menu} for compatibility with \code{shinydashboard})
+#' @description Create a dropdown menu with additional UI elements.
+#' @param ... UI elements to include within the dropdown menu.
+#' @param type Type of the displayed items.
+#' @return A dropdown menu that can be passed to \code{\link[semantic.dashboard]{dashboardHeader}}
+#' @export
+dropdown_menu <- function(..., type = "messages"){
+  icon <- if (type == "messages") {
```
Comment (marekrogala) [inline R/uidropdown_menu.R]: 1. instead of a series of if-elses, how about making a named list `icon_for_dropdown_menu_type` and do a simple lookup here?
2. you could also add a verification step here that type is in `names(icon_for_dropdown_menu_type)`

File: R/uidropdown_menu.R [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+#' Create a task item (alias for \code{task_item} for compatibility with \code{shinydashboard})
+#' @description Create a task item.
+#' @param text Progress bar label.
+#' @param value Progress bar value.
+#' @param color Color of the task item. One of \code{c("", "red", "orange", "yellow", "olive", "green", "teal", "blue", "violet", "purple", "pink", "brown", "grey", "black")}
+#' @return A task item that can be passed to \code{\link[semantic.dashboard]{dropdownMenu}}
+#' @export
+task_item <- function(text, value, color = ""){
+  if (!is.numeric(value) || (value < MIN_PROGRESS_VALUE) || (value > MAX_PROGRESS_VALUE))
+    stop(paste("'value' must be between", MIN_PROGRESS_VALUE, "and", MAX_PROGRESS_VALUE))
```
Comment (marekrogala) [inline R/uidropdown_menu.R]: 1. missing { } around this 
2. why do we use `stop` here, when we used `warning` in other similar cases?

File: R/uidropdown_menu.R [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  verify_value_allowed("color", c("", ALLOWED_COLORS))
+  shiny::tags$div(class = "item", style = "min-width:200px;",
```
Comment (marekrogala) [inline R/uidropdown_menu.R]: nit: missing space after `:` in css

File: R/uidropdown_menu.R [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+#' Create a notification item (alias for \code{notification_item} for compatibility with \code{shinydashboard})
+#' @description Create a notification item.
+#' @param text Text of the message.
+#' @param icon Additional icon.
+#' @param color Color of the notification item. One of \code{c("", "red", "orange", "yellow", "olive", "green", "teal", "blue", "violet", "purple", "pink", "brown", "grey", "black")}
+#' @return A notification item that can be passed to \code{\link[semantic.dashboard]{dropdownMenu}}
+#' @export
+notification_item <- function(text, icon = "warning", color = ""){
+  verify_value_allowed("color", c("", ALLOWED_COLORS))
```
Comment (marekrogala) [inline R/uidropdown_menu.R]: this step is repeated in multiple places, please extract it to a function like `verify_color_allowed("color")`.
BTW it would be nicer if we could call these functions like `verify_color_allowed(color)` (and still get these nice error messages containing argument's name) --- but this is out of scope of this PR

File: R/uidropdown_menu_output.R [hunk 7]
```diff
@@ -0,0 +1,28 @@
+#' Create a dropdown menu output (alias for \code{value_box_output} for compatibility with \code{shinydashboard})
```
Comment (marekrogala) [inline R/uidropdown_menu_output.R]: 1. why is this file called `uidropdown_menu_output` instead of simply `dropdown_menu_output` as the main function in it?
2. are you sure this is the right documentation ? `(alias for \code{value_box_output} for compatibility with \code{shinydashboard})`. BTW you don't have to copy paste, RStudio has an option to "insert roxygen skeleton" (Ctrl Alt Shift R)

File: R/uidropdown_menu_output.R [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+#' @description UI-side function for dynamic valueBox.
```
Comment (marekrogala) [inline R/uidropdown_menu_output.R]: are you sure that this is the right documentation? please make sure that we have the right documentation for each function we're adding here

File: R/uidropdown_menu_output.R [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+#' Create a dropdown menu output (alias for \code{dropdown_menu_output} for compatibility with \code{shinydashboard})
+#' @description Server-side function for dynamic dropdownMenu.
+#' @param expr dropdownMenu.
+#' @param env The environment in which to evaluate expr.
+#' @param quoted Is expr a quoted expression (with \code{quote()})? This is useful if you want to save an expression in a variable.
+#' @return A dynamic dropdown menu that can be assign to output.
+#' @export
+render_dropdown_menu <- function(expr, env = parent.frame(), quoted = FALSE){
+  vbox_fun <- shiny::exprToFunction(expr, env, quoted)
```
Comment (marekrogala) [inline R/uidropdown_menu_output.R]: why `vbox_fun`?

File: R/uivalue_box.R [hunk 10]
```diff
@@ -19,3 +19,7 @@ value_box <- function(title, value, icon = NULL, color = "blue", width = 5){
 #' @describeIn value_box Create a valueBox (alias for \code{value_box} for compatibility with \code{shinydashboard})
 #' @export
 valueBox <- value_box
+
+#' @describeIn value_box Create a valueBox (alias for \code{value_box} for compatibility with \code{shinydashboard})
+#' @export
+infoBox <- value_box
```
Comment (marekrogala) [inline R/uivalue_box.R]: :+1:

File: R/uivalue_box_output.R [hunk 11]
```diff
@@ -12,18 +12,26 @@ value_box_output <- function(outputId, width = 5){
...
+#' @export
+infoBoxOutput <- value_box_output
+
+#' Create a value box output (alias for \code{render_value_box} for compatibility with \code{shinydashboard})
 #' @description Server-side function for dynamic valueBox.
 #' @param expr ValueBox.
 #' @param env The environment in which to evaluate expr.
 #' @param quoted Is expr a quoted expression (with \code{quote()})? This is useful if you want to save an expression in a variable.
-#' @return A dynamic valueBox that can be passed to \code{\link[semantic.dashboard]{valueBoxOutput}}
+#' @return A dynamic valueBox that can be assign to output.
```
Comment (marekrogala) [inline R/uivalue_box_output.R]: *be assigned

File: examples/app.R [hunk 12]
```diff
@@ -7,12 +7,16 @@ if(interactive()){
   library(DT)
 
   ui <- dashboardPage(
-    dashboardHeader(color = "black"),
+    dashboardHeader(color = "black",
+                    dropdownManuOutput("dropdown1"),
```
Comment (marekrogala) [inline examples/app.R]: typo, please make sure that all exmples work correctly

File: NAMESPACE [hunk 13]
```diff
@@ -10,17 +10,32 @@ export(dashboard_body)
 export(dashboard_header)
 export(dashboard_page)
 export(dashboard_sidebar)
+export(dropdownManuOutput)
```
[the remark this answers is not in the document]
Comment (maju116) [reply NAMESPACE]: done

File: R/semantic_dashboard.R (hunk 2 above)
Comment (maju116) [reply R/semantic_dashboard.R]: done

File: R/uidropdown_menu.R (hunk 3 above)
Comment (maju116) [reply R/uidropdown_menu.R]: For now this code is a copy of `shinydashboard::dropdownMenu`, where user can only specify 3 types of items. The truth is that in pour function user can add any item to dropdown with `div(class = "item", ...)` so maybe we should the user a free will for the icon and later remove `type` agr? @marekrogala

File: R/uidropdown_menu.R (hunk 4 above)
Comment (maju116) [reply R/uidropdown_menu.R]: 1. what for, it's only one line ? You didn't want it in previous rev
2. changed
It's a similar case, for now we accept only percents for values but in the future `max_progres_value` parameters...

File: R/uidropdown_menu.R (hunk 5 above)
Comment (maju116) [reply R/uidropdown_menu.R]: done

File: R/uidropdown_menu.R (hunk 6 above)
Comment (maju116) [reply R/uidropdown_menu.R]: For now it will stay, there will be some changes in colors in the future...

File: R/uidropdown_menu_output.R (hunk 7 above)
Comment (maju116) [reply R/uidropdown_menu_output.R]: names changed

File: R/uidropdown_menu_output.R (hunk 8 above)
Comment (maju116) [reply R/uidropdown_menu_output.R]: changed

File: R/uidropdown_menu_output.R (hunk 9 above)
Comment (maju116) [reply R/uidropdown_menu_output.R]: Thats `shiny` nomenclature

File: R/uivalue_box_output.R (hunk 11 above)
Comment (maju116) [reply R/uivalue_box_output.R]: changed

File: examples/app.R (hunk 12 above)
Comment (maju116) [reply examples/app.R]: done

File: tests/testthat/test-names.R (hunk 1 above)
Comment (maju116) [reply tests/testthat/test-names.R]: done

File: R/uidropdown_menu.R (hunk 3 above)
Comment (marekrogala) [reply R/uidropdown_menu.R]: @maju116 this would be nice, but how do we solve compatibility with shinydashboard values in this case, so that user doesn't have to modify type when switching to semantic.dashboard?

File: R/uidropdown_menu.R (hunk 4 above)
Comment (marekrogala) [reply R/uidropdown_menu.R]: 1. This comes from our style guide, although I think it was not clear enough (only from examples). I updated the style guide to be more explicit: https://github.com/Appsilon/styleguide/blob/master/R-style-guide/R.md . See also examples in Hadley's version of the style guide: http://adv-r.had.co.nz/Style.html

File: R/uidropdown_menu.R (hunk 6 above)
Comment (marekrogala) [reply R/uidropdown_menu.R]: ok, as discussed offline

File: R/uidropdown_menu_output.R (hunk 9 above)
Comment (marekrogala) [reply R/uidropdown_menu_output.R]: vbox sounds like value box, but we're in a dropdown menu
we don't have to blindly follow shiny's names
why not simply `fun`?

File: R/uidropdown_menu.R (hunk 3 above)
Comment (maju116) [reply R/uidropdown_menu.R]: @marekrogala like you said "we don't have to have 100% compatibility". I think this is a great case when we can do something better/more than `shinydashboard`. @filipstachura what do you think ?

File: R/uidropdown_menu.R (hunk 4 above)
Comment (maju116) [reply R/uidropdown_menu.R]: changed

File: R/uidropdown_menu_output.R (hunk 9 above)
Comment (maju116) [reply R/uidropdown_menu_output.R]: changed to `fun`

File: R/uidropdown_menu.R (hunk 3 above)
Comment (marekrogala) [reply R/uidropdown_menu.R]: I see. I think this is a place where it's easy for us to maintain compatibility, so I'd do that -- it is large value if it's easy to switch.

That said, I think you're right that this is something we can do better than shinydashboard.

I think we can combine both of the above: if keep the `type` argument, but at the same time add an `icon` argument that is directly used. E.g. if you provide icon, we'll use icon, otherwise we'll select the icon based on type. WDYT?

File: R/dropdown_menu.R [hunk 14]
```diff
@@ -0,0 +1,83 @@
+#' Create a dropdown menu (alias for \code{dropdown_menu} for compatibility with \code{shinydashboard})
```
Comment (marekrogala) [inline R/dropdown_menu.R]: still an error in documentation: unneeded `(alias for \code{dropdown_menu} for compatibility with \code{shinydashboard})` - this is not an alias

File: R/dropdown_menu.R [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+                     shiny::tags$div(class = "menu", ...),
+                     shiny::tags$script(dropdown_menu_js),
+                     shiny::tags$script(progress_bar_js))
+}
+
+#' @describeIn dropdown_menu Create a dropdown menu (alias for \code{dropdown_menu} for compatibility with \code{shinydashboard})
+#' @export
+dropdownMenu <- dropdown_menu
+
+#' Create a message item (alias for \code{message_item} for compatibility with \code{shinydashboard})
```
Comment (marekrogala) [inline R/dropdown_menu.R]: ditto

File: R/dropdown_menu.R [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+  shiny::tags$div(class = paste("ui icon message", color), shiny::tags$i(class = paste("small", icon, "icon")),
+                  shiny::tags$div(class = "content",
+                                  shiny::tags$div(class = "header", from), message))
+}
+
+#' @describeIn message_item Create a message item (alias for \code{message_item} for compatibility with \code{shinydashboard})
+#' @export
+messageItem <- message_item
+
+#' Create a task item (alias for \code{task_item} for compatibility with \code{shinydashboard})
```
Comment (marekrogala) [inline R/dropdown_menu.R]: ditto

File: R/dropdown_menu.R [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+                                  `data-percent` = value, `data-total` = 100,
+                                  shiny::tags$div(class = "bar", shiny::tags$div(class = "progress")),
+                                  shiny::tags$div(class = "label", text)))
+}
+
+#' @describeIn task_item Create a task item (alias for \code{taks_item} for compatibility with \code{shinydashboard})
+#' @export
+taskItem <- task_item
+
+#' Create a notification item (alias for \code{notification_item} for compatibility with \code{shinydashboard})
```
Comment (marekrogala) [inline R/dropdown_menu.R]: ditto

File: R/dropdown_menu_output.R [hunk 18]
```diff
@@ -0,0 +1,28 @@
+#' Create a dropdown menu output (alias for \code{dropdown_menu_output} for compatibility with \code{shinydashboard})
```
Comment (marekrogala) [inline R/dropdown_menu_output.R]: ditto

File: R/dropdown_menu_output.R [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+#' @export
+dropdown_menu_output <- function(outputId){
+  shiny::uiOutput(outputId)
+}
+
+#' @describeIn dropdown_menu_output Create a value box output (alias for \code{dropdown_menu_output} for compatibility with \code{shinydashboard})
+#' @export
+dropdownMenuOutput <- dropdown_menu_output
+
+#' Create a dropdown menu output (alias for \code{render_dropdown_menu} for compatibility with \code{shinydashboard})
```
Comment (marekrogala) [inline R/dropdown_menu_output.R]: ditto

File: R/value_box_output.R [hunk 20]
```diff
@@ -0,0 +1,37 @@
+#' Create a value box output (alias for \code{value_box_output} for compatibility with \code{shinydashboard})
```
Comment (marekrogala) [inline R/value_box_output.R]: ditto

File: R/value_box_output.R [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+
+#' @describeIn value_box_output Create a valueBox output (alias for \code{value_box_output} for compatibility with \code{shinydashboard})
+#' @export
+valueBoxOutput <- value_box_output
+
+#' @describeIn value_box_output Create a valueBox output (alias for \code{value_box_output} for compatibility with \code{shinydashboard})
+#' @export
+infoBoxOutput <- value_box_output
+
+#' Create a value box output (alias for \code{render_value_box} for compatibility with \code{shinydashboard})
```
Comment (marekrogala) [inline R/value_box_output.R]: ditto

File: R/uidropdown_menu.R (hunk 3 above)
Comment (maju116) [reply R/uidropdown_menu.R]: good idea

File: R/value_box_output.R (hunk 20 above)
Comment (maju116) [reply R/value_box_output.R]: done

File: R/dropdown_menu_output.R (hunk 19 above)
Comment (maju116) [reply R/dropdown_menu_output.R]: done

File: R/dropdown_menu_output.R (hunk 18 above)
Comment (maju116) [reply R/dropdown_menu_output.R]: done

File: R/dropdown_menu.R (hunk 17 above)
Comment (maju116) [reply R/dropdown_menu.R]: done

File: R/dropdown_menu.R (hunk 16 above)
Comment (maju116) [reply R/dropdown_menu.R]: done

File: R/dropdown_menu.R (hunk 15 above)
Comment (maju116) [reply R/dropdown_menu.R]: done

File: R/dropdown_menu.R (hunk 14 above)
Comment (maju116) [reply R/dropdown_menu.R]: done

File: R/dropdown_menu.R [hunk 22]
```diff
@@ -1,16 +1,15 @@
...
-  icon <- if (type == "messages") {
-    icon("mail")
-  } else if (type == "notifications"){
-    icon("warning sign")
+dropdown_menu <- function(..., type = "messages", icon = NULL){
+  icon <- if (!is.null(icon)) {
+    icon
   } else {
-    icon("tasks")
+    icon(DROPDOWN_MENU_ICONS[type])
```
Comment (marekrogala) [inline R/dropdown_menu.R]: oopss, I can see a conflict of names here -- are you sure this works correctly?

Comment (maju116) [reply R/dropdown_menu.R]: There was an argument error in app.R, fixed

**Comment by marekrogala** [comment]
LGTM!
